File: src/probe.ts
// f/vless/probe.ts

import type { ProbeResult } from "./types";
import { Semaphore } from "./semaphore";

export async function tcpProbeAll(
  hosts: Array<{ host: string; port: number; ip: string | null }>,
  maxConcurrent: number,
  timeoutMs: number
): Promise<Map<string, ProbeResult>> {
  const results = new Map<string, ProbeResult>();
  const sem = new Semaphore(maxConcurrent);

  await Promise.allSettled(
    hosts.map(async ({ host, port, ip }) => {
      await sem.acquire();
      try {
        const target = ip ?? host;
        const key = `${host}:${port}`;
        const start = performance.now();
        const result = await tcpProbe(target, port, timeoutMs);
        const latencyMs = result.reachable ? Math.round(performance.now() - start) : null;
        results.set(key, { ...result, latencyMs });
      } finally {
        sem.release();
      }
    })
  );

  return results;
}

function tcpProbe(hostname: string, port: number, timeoutMs: number): Promise<ProbeResult> {
  return new Promise((resolve) => {
    let settled = false;
    const mark = (result: ProbeResult) => {
      if (!settled) {
        settled = true;
        resolve(result);
      }
    };

    const timer = setTimeout(() => {
      mark({ reachable: false, latencyMs: null, failureType: "timeout" });
    }, timeoutMs);

    const socket = Bun.connect({
      hostname,
      port,
      socket: {
        data() {},
        open(socket) {
          clearTimeout(timer);
          mark({ reachable: true, latencyMs: null, failureType: null });
          socket.end();
        },
        connectError(_socket, error) {
          clearTimeout(timer);
          const msg = String(error);
          if (msg.includes("ECONNREFUSED")) {
            mark({ reachable: false, latencyMs: null, failureType: "refused" });
          } else {
            mark({ reachable: false, latencyMs: null, failureType: "error" });
          }
        },
        error(_socket, _error) {
          clearTimeout(timer);
          mark({ reachable: false, latencyMs: null, failureType: "error" });
        },
        close(_socket, _error) {
          clearTimeout(timer);
          if (!settled) {
            mark({ reachable: false, latencyMs: null, failureType: "error" });
          }
        },
        timeout(socket) {
          clearTimeout(timer);
          mark({ reachable: false, latencyMs: null, failureType: "timeout" });
          socket.end();
        },
        end() {
          clearTimeout(timer);
          if (!settled) {
            mark({ reachable: false, latencyMs: null, failureType: "error" });
          }
        },
      },
    });

    socket.catch(() => {
      clearTimeout(timer);
      mark({ reachable: false, latencyMs: null, failureType: "error" });
    });
  });
}

File: internal/probe/probe.go
package probe

import (
	"context"
	"fmt"
	"net"
	"sync"
	"time"
)

type ProbeResult struct {
	Reachable   bool
	LatencyMs   int64
	FailureType string // "refused", "timeout", "error"
}

type HostSpec struct {
	Host string
	IP   string
	Port int
}

func TCPProbeAll(ctx context.Context, hosts []HostSpec, maxConcurrent int, timeout time.Duration) map[string]*ProbeResult {
	results := make(map[string]*ProbeResult, len(hosts))
	var mu sync.Mutex
	sem := make(chan struct{}, maxConcurrent)
	var wg sync.WaitGroup

	for _, h := range hosts {
		wg.Add(1)
		sem <- struct{}{}
		go func(host, ip string, port int) {
			defer wg.Done()
			defer func() { <-sem }()
			select {
			case <-ctx.Done():
				return
			default:
			}
			key := fmt.Sprintf("%s:%d", host, port)
			target := ip
			if target == "" {
				target = host
			}
			result := tcpProbe(ctx, target, port, timeout)
			mu.Lock()
			results[key] = result
			mu.Unlock()
		}(h.Host, h.IP, h.Port)
	}

	wg.Wait()
	return results
}

func tcpProbe(ctx context.Context, host string, port int, timeout time.Duration) *ProbeResult {
	addr := fmt.Sprintf("%s:%d", host, port)
	start := time.Now()
	d := net.Dialer{Timeout: timeout}
	conn, err := d.DialContext(ctx, "tcp", addr)
	if err != nil {
		ft := "error"
		if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
			ft = "timeout"
		} else {
			ft = "refused"
		}
		return &ProbeResult{Reachable: false, FailureType: ft}
	}
	conn.Close()
	return &ProbeResult{Reachable: true, LatencyMs: time.Since(start).Milliseconds()}
}
